### ISSUE TITLE
Usage with hyper's RequestBuilder

### ISSUE THREAD

**Issue by anowell** [cybergeek94/multipart#23]
I'm finding it quite challenging to retrofit multipart into a client library that was built using hyper's `RequestBuilder`. I've been trying to implement `HttpRequest` and `HttpStream` for a wrapper around hyper's `RequestBuilder` (I can't see an obvious way to implement it directly against RequestBuilder), and perhaps my Rust is still just inadequate, but I keep hitting walls. I've worked around a few limitations of RequestBuilder by handling it very much like SizedRequest in this library (buffer input to a Vec<u8> and defer setting headers until `finish` is called since RequestBuilder consumes self to set headers).

However, I can't seem to call RequestBuilder's `body` method inside `finish` given the trait's lifetimes: RequestBuilder only borrows the buffer, and (if I understand correctly), I can't ensure that self (which owns the buffer) outlives the response (I'm not sure I grok why it even needs to outlive the response, but the borrow checker has defeated me for today.)

I thought I'd reach out and see if there was an obvious answer that I've been overlooking all along. Are there any examples of multipart used in conjunction with hyper's RequestBuilder? Is this even in-scope for this lib?

```rust
pub struct MultipartBuilder<'a> {
    inner: RequestBuilder<'a, Url>,
    boundary: String,
    buffer: Vec<u8>,
}

impl <'a> HttpRequest for MultipartBuilder<'a>  {
    type Stream = Self;
    type Error = HyperError;

    fn apply_headers(&mut self, boundary: &str, content_len: Option<u64>) -> bool {
        // Can't actually set headers, because RequestBuilder's header method consumes self
        self.boundary = boundary.into();
        true
    }
    fn open_stream(mut self) -> Result<Self::Stream, Self::Error> {
        // Can't actually get RequestBuilder's stream, so we'll buffer it and handle it in `finish`
        self.buffer.clear();
        Ok(self)
    }

}

impl <'a> HttpStream for MultipartBuilder<'a>  {
    type Request = Self;
    type Response = Response;
    type Error = HyperError;

    fn finish(mut self) -> Result<Self::Response, Self::Error> {
        self.inner
            .header(ContentType(Mime(
                TopLevel::Multipart, SubLevel::Ext("form-data".into()),
                vec![(Attr::Ext("boundary".into()), Value::Ext(self.boundary))]
            )))
            .header(ContentLength(self.buffer.len() as u64))
            .body(&*self.buffer)  // BAD: &*self.buffer doesn't live long enough
            .send()
    }
}
```

**Comment by cybergeek94** [comment]
So `RequestBuilder` has a lifetime `'a`, which, as you guessed correctly, means it only borrows the buffer passed to `.body()`, so it requires that that buffer *outlive* `'a`.

There's a couple possible ways you could fix this. I think you should be able to lift `self.inner` to a local `let` binding, which should hint to the compiler that it won't outlive `finish()`. It may necessitate lifting the borrow of `self.buffer` to the same level:

```rust
let buffer = &*self.buffer;
let inner = self.inner;
inner.header(...)
    .header(...)
    .body(buffer)
    .send();
```

The other would be to delay creating the `RequestBuilder`, only storing the `Url` and then using it to construct `RequestBuilder` in `finish()`. That way you don't have a lifetime to bind to `MultipartBuilder`.

The client-side of `multipart` wasn't really designed with `RequestBuilder` in mind. That feature came a little while after principle development and `RequestBuilder` didn't seem flexible enough to support important features in `multipart`, like writing the request fields directly to the stream as they come in.

Is there anything wrong with using `multipart` as it was designed? The `hyper::client::Request` API really is much more flexible.

**Comment by anowell** [comment]
Thanks for taking the time to suggest some changes on my end. Unfortunately lifting `buffer` and `inner` didn't resolve it. I suspect it's because the builder pattern is consuming `self` at every step so the compiler seems to believe the borrowed buffer is part of the object being returned. I'll chew on your other suggestion a bit more, although in the context of an API client with lots of non-multipart requests, it feels a bit strange to have a very different mechanism for injecting other headers (like Auth) into multipart requests.

I agree that hyper's `RequestBuilder` isn't nearly as flexible as `Request`. I did start down the route of ripping out all usage of `RequestBuilder`, and `multipart` is great for working with `Request`, but `Request` just felt too low-level for what I want to accomplish. Perhaps I overcomplicated it, but I quickly found myself basically reinventing `RequestBuilder`: shamelessly copying from `RequestBuilder` to configure/share a connection pool, handle timeouts, follow redirects, think about chunked vs sized requests and set content-length myself (for non-multipart requests), and then expose all that in a ergonomic module that I could use throughout the rest of the library. I felt like it was a lot of overhead just to add support for multipart requests.

`RequestBuilder` provides a really simple abstraction that frees me from thinking of all of that. While it may be out of scope for this this project, it seems like there is a place for combining that level of simplicity with the ability to make multipart requests. I'm just trying to decide if I feel strongly enough about it to build it.
